Rate Reminder implementation

pritam-bs opened the pull request:
- Both default and custom configuration strategies
- Documentation for Rate Reminder usage
- Example for Rate Reminder
- Some clean-up in Message and Version-Control
- Fix the template for enabling the test mode

pritam-bs committed:
- Rate Reminder implementation
- Both default and custom configuration strategies
- Documentation for Rate Reminder usage
- Example for Rate Reminder
- Some clean-up in Message and Version-Control
- Fix the template for enabling the test mode

hassan-saleh-ml commented on lib/src/models/rate_reminder.dart:
```diff
@@ -1,28 +1,52 @@
-  final String? link;
+class NStackRateReminder {
+  final int id;
+  final int points_to_trigger;
+  final int days_delay_on_skip;
```
I would avoid having underscores in the variable name, @nivisi shouldn't the analyzer fail on this?

Are we not using monstarlab_lints package with NStack?

pritam-bs replied:
Yeah, monstarlab_lints is used. Should I need to update the package?

nivisi replied:
@hassan-saleh-ml I don't think there's a linter rule for this. But we should follow `lowerCamelCase` @pritam-bs

nivisi replied:
Actually there is a rule for this! [non_constant_identifier_names](https://dart.dev/tools/linter-rules/non_constant_identifier_names). Good catch. I will add it to our linter rules.

hassan-saleh-ml commented on lib/src/models/rate_reminder.dart:
```diff
@@ -1,28 +1,52 @@
+  }
+}
+
+class Localization {
```
Maybe rename this to something specific to rate reminder as Localization might be used by Flutter or other packages.

hassan-saleh-ml commented on lib/src/repository/nstack_repository.dart:
```diff
@@ -203,4 +170,83 @@ class NStackRepository {
+    var mutableHeaders = {..._headers};
+    mutableHeaders['Accept-Language'] = acceptHeader;
+    try {
+      final rateReminderResponse = await http.get(
+        Uri.parse(
+          '$_baseUrl/notify/rate_reminder_v2?guid=${appInfoData.guid}',
+        ),
+        headers: mutableHeaders,
+      );
```
I remember @BucekJiri said that we don't have to fetch anything from the API except for app open, @pritam-bs can't we get the info here from app open response since we already fetched this information?

pritam-bs replied:
Open app API does not provide the rate reminder information.
https://nstack-io.github.io/docs/docs/features/rate-reminder.html?sidebarScroll=0

hassan-saleh-ml commented on lib/src/repository/nstack_repository.dart:
```diff
@@ -203,4 +170,83 @@ class NStackRepository {
+
+      final result = json.decode(rateReminderResponse.body);
+
+      if (rateReminderResponse.statusCode == 445) {
```
This is a very specific error code, are there any error codes we should handle?

Shouldn't all `4xx` status codes be handled?

pritam-bs replied:
This log message mainly informs the developer how many points must be triggered before the app can show the rate reminder alert.
the other errors will be logged in the catch block.

nivisi replied:
@pritam-bs can we then add a comment right in the code for future generations?

hassan-saleh-ml commented on lib/src/sdk/nstack_features.dart:
```diff
@@ -8,6 +10,14 @@ typedef OnVersionUpdateNotification = void Function(
 )?;
 
+typedef OnRateReminder = void Function(
+  NStackRateReminder? rateReminderInfo,
```
I don't think this should be nullable, when `onRateReminder` we should expect the info to never be null.

Same for `NStackRateReminderAnswer? rateReminderAnswer,`

hassan-saleh-ml replied:
Also the method in the type def itself should not be nullable, for example:

```dart
typedef OnRateReminderAnswered = void Function(
  NStackRateReminderAnswer rateReminderAnswer,
);
```

Then in the class you can make it nullable:

```dart
  final OnRateReminder? onRateReminder;
  final OnRateReminderAnswered? onRateReminderAnswered;
```

pritam-bs replied:
The `NStackRateReminder` is nullable for the `onRateReminder` callback when the points are less than the certain threshold set in the console, and the devs can make a conditional check if needed.
But `NStackRateReminderAnswer` should not be nullable; I will fix that.

hassan-saleh-ml replied:
@pritam-bs if we recieve null from BE should we display the reminder? If we should not display it then we should not call the callback in the first place and it can be nun nullable

pritam-bs replied:
The intention was to enable developers to check the `NStackRateReminder` value and decide whether to do anything when it is null. Anyway, change it to not nullable.

pritam-bs committed:
- snake case to camel case
- Remane Localization for Rate Reminder
- method in the type def made not nullable

nivisi commented on lib/src/sdk/nstack_features.dart:
```diff
@@ -1,12 +1,22 @@
-)?;
+);
+
+typedef OnRateReminder = void Function(
+  NStackRateReminder? rateReminderInfo,
+);
+
+typedef OnRateReminderAnswered = void Function(
+  NStackRateReminderAnswer rateReminderAnswer,
+);
```
On another topic for typedefs. There is no written naming convention available (based on a quick research I just did), but usually these callbacks must be named differently. Here are a few examples from the Flutter repo (see below).

I recommend to follow these rules:

```suggestion
typedef RateReminderCallback = void Function(
  NStackRateReminder data,
);

typedef RateReminderAnswerCallback = void Function(
  NStackRateReminderAnswer answer,
);
```

---

## Examples

```dart
typedef ValueGetter<T> = T Function();

/// Signature for callbacks that filter an iterable.
typedef IterableFilter<T> = Iterable<T> Function(Iterable<T> input);

/// Signature of callbacks that have no arguments and return no data, but that
/// return a [Future] to indicate when their work is complete.
///
/// See also:
///
///  * [VoidCallback], a synchronous version of this signature.
///  * [AsyncValueGetter], a signature for asynchronous getters.
///  * [AsyncValueSetter], a signature for asynchronous setters.
typedef AsyncCallback = Future<void> Function();

/// Signature for callbacks that report that a value has been set and return a
/// [Future] that completes when the value has been saved.
///
/// See also:
///
///  * [ValueSetter], a synchronous version of this signature.
///  * [AsyncValueGetter], the getter equivalent of this signature.
typedef AsyncValueSetter<T> = Future<void> Function(T value);
```

```dart
typedef GestureDragEndCallback = void Function(DragEndDetails details);

/// Signature for when the pointer that previously triggered a
/// [GestureDragDownCallback] did not complete.
///
/// Used by [DragGestureRecognizer.onCancel].
typedef GestureDragCancelCallback = void Function();

/// Signature for a function that builds a [VelocityTracker].
///
/// Used by [DragGestureRecognizer.velocityTrackerBuilder].
typedef GestureVelocityTrackerBuilder = VelocityTracker Function(PointerEvent event);
```

nivisi commented on lib/src/sdk/widgets/nstack_rate_reminder_widget.dart:
```diff
@@ -0,0 +1,180 @@
+          'If onRateReminder is null, then onAnswered must not be null.',
+        );
+
+  final Widget? child;
```
Put `child` widgets last in the property list. Again, this is a convention that majority of Flutter widgets follow (and we do follow it as well). Same for the constructor parameter list

There's a related rule for this: https://dart.dev/tools/linter-rules/sort_child_properties_last

nivisi commented on lib/src/sdk/widgets/nstack_rate_reminder_widget.dart:
```diff
@@ -0,0 +1,180 @@
+    this.onAnswered,
+  }) : assert(
+          onRateReminder != null || onAnswered != null,
+          'If onRateReminder is null, then onAnswered must not be null.',
```
```suggestion
          'Either onRateReminder or onAnswered must be provided.',
```

nivisi commented on lib/src/sdk/widgets/nstack_rate_reminder_widget.dart:
```diff
@@ -0,0 +1,180 @@
+        );
+
+  final Widget? child;
+  final void Function(NStackRateReminder?)? onRateReminder;
+  final void Function(NStackRateReminderAnswer)? onAnswered;
```
Use our `typedef`s here

nivisi commented on lib/src/sdk/widgets/nstack_rate_reminder_widget.dart:
```diff
@@ -0,0 +1,180 @@
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+
+    WidgetsBinding.instance.addPostFrameCallback(
+      (timeStamp) async {
+        final rateReminderInfo =
+            await context.nstack.rateReminders.getRateReminderInfo(
+          defaultLocale: Localizations.localeOf(context),
+        );
+        _onRateReminder(rateReminderInfo);
+      },
+    );
```
We don't need a post frame callback in `didChangeDependencies`, it's only required in `initState`.

Although are we sure this needs to be done in `didChangeDependencies`? It might get executed quite frequently here

nivisi commented on lib/src/sdk/widgets/nstack_rate_reminder_widget.dart:
```diff
@@ -0,0 +1,180 @@
+      rateReminder.localization?.laterBtn ?? _laterButtonTitleFallback,
+    );
+
+    Future<void> buttonAction(
```
Let's move this out of the `build` method

nivisi commented on lib/templates/nstack_template.txt:
```diff
@@ -199,18 +249,20 @@ class NStackLocalizationAsset {
     required Widget child,
     AppOpenPlatform? platformOverride,
     VoidCallback? onComplete,
+    bool? testMode,
   }) : super(
           key: key,
           child: child,
           platformOverride: platformOverride,
           onComplete: onComplete,
           config: config,
+          testMode: testMode ?? false,
```
```suggestion
    bool testMode = false,
  }) : super(
          key: key,
          child: child,
          platformOverride: platformOverride,
          onComplete: onComplete,
          config: config,
          testMode: testMode,
```

nivisi commented on lib/src/sdk/widgets/nstack_rate_reminder_widget.dart:
```diff
@@ -0,0 +1,180 @@
+  final void Function(NStackRateReminderAnswer)? onAnswered;
+
+  @override
+  State<StatefulWidget> createState() => _NStackRateReminderWidgetSate();
+}
+
+class _NStackRateReminderWidgetSate extends State<NStackRateReminderWidget> {
```
```suggestion
  State<StatefulWidget> createState() => _NStackRateReminderWidgetState();
}

class _NStackRateReminderWidgetState extends State<NStackRateReminderWidget> {
```

pritam-bs committed:
- Clean up to follow Flutter convention

pritam-bs committed:
- sort child properties last
